src/server/json_family.cc

```diff
@@ -692,8 +702,135 @@ OpResult<vector<OptSizeT>> OpArrTrim(const OpArgs& op_args, string_view key, str
   return vec;
 }
 
+OpResult<vector<OptLong>> OpArrIndex(const OpArgs& op_args, string_view key,
+                                     JsonExpression expression, const json& search_val,
+                                     int start_index, int end_index) {
+  OpResult<json> result = GetJson(op_args, key);
+  if (!result) {
+    return result.status();
+  }
+
+  vector<OptLong> vec;
+  auto cb = [&](const string_view& path, const json& val) {
+    size_t search_start_index;
+    size_t search_end_index;
+    if (!val.is_array()) {
+      vec.emplace_back(nullopt);
+      return;
+    }
+
+    if (val.empty()) {
+      vec.emplace_back(-1);
+      return;
+    }
+
+    if (start_index < 0) {
+      int idx = val.size() + start_index;
+      if (idx < 0) {
+        idx %= val.size();
+      }
+
+      search_start_index = abs(idx);
+    } else {
+      search_start_index = start_index;
+    }
+
+    if (end_index <= 0) {
+      if (end_index == 0 || end_index == -1) {
```

Comment:
maybe I can help with combining those if's if you write here what is the expected mapping for negative values.

Reply:
Negative index values are treated by rounding the index to the array's start and end.

Reply:
Please add a comment at line 728 with example on how say -2 is mapped when val.size() is 9.
and how -11 should map.

I think it's possible to improve the code. at line 728:
```cpp
if (start_index < 0) {
  start_index %= val.size();  // start_index stays negative but greater than -val.size().
  start_index += val.size();  // 0 < start_index < val.size()
}
```
no need search_start_index. Similarly with end_index.